libraries/vmmanager/fizzy/fizzy_vm_backend.cpp

```diff
@@ -0,0 +1,269 @@
+
+   FizzyValueType invoke_system_call_arg_types[] = {FizzyValueTypeI32, FizzyValueTypeI32, FizzyValueTypeI32, FizzyValueTypeI32, FizzyValueTypeI32};
+   size_t invoke_system_call_num_args = 5;
+   FizzyExternalFunction invoke_system_call_fn = {{ FizzyValueTypeVoid, invoke_system_call_arg_types, invoke_system_call_num_args }, invoke_system_call, this };
+
+   size_t num_host_funcs = 2;
+   FizzyImportedFunction host_funcs[] = {{"env", "invoke_thunk", invoke_thunk_fn}, {"env", "invoke_system_call", invoke_system_call_fn}};
+
+   FizzyError fizzy_err;
+
+   // TODO:  Make memory_pages_limit configurable
+   size_t memory_pages_limit = 512;     // Number of 64k pages allowed to allocate
+
+   KOINOS_ASSERT( _instance == nullptr, runner_state_exception, "_instance was unexpectedly non-null" );
+   _instance = fizzy_resolve_instantiate(_module, host_funcs, num_host_funcs, nullptr, nullptr, nullptr, 0, memory_pages_limit, &fizzy_err);
+   if( _instance == nullptr )
+   {
+      std::string error_code = fizzy_error_code_name( fizzy_err.code );
+      std::string error_message = fizzy_err.message;
+      KOINOS_THROW( module_instantiate_exception, "Could not instantiate module - ${code}: ${msg}", ("code", error_code)("msg", error_message) );
+   }
+}
+
+FizzyExecutionResult fizzy_runner::_invoke_thunk( const FizzyValue* args, FizzyExecutionContext* fizzy_context ) noexcept
+{
+   FizzyExecutionResult result;
+   result.has_value = false;
+   result.value.i64 = 0;
+
+   _exception = std::exception_ptr();
+
+   try
+   {
+      uint32_t tid = args[0].i32;
+      uint32_t ret_len = args[2].i32;
+      char* ret_ptr = resolve_ptr(_instance, args[1].i32, ret_len);
+      uint32_t arg_len = args[4].i32;
+      const char* arg_ptr = resolve_ptr(_instance, args[3].i32, arg_len);
+
+      KOINOS_ASSERT( ret_ptr != nullptr, wasm_memory_exception, "Invalid ret_ptr in invoke_thunk()" );
+      KOINOS_ASSERT( arg_ptr != nullptr, wasm_memory_exception, "Invalid arg_ptr in invoke_thunk()" );
+
+      KOINOS_ASSERT( _ctx._api_handler != nullptr, null_argument_exception, "_api_handler was unexpectedly null pointer" );
+      _ctx._api_handler->invoke_thunk( tid, ret_ptr, ret_len, arg_ptr, arg_len );
+   }
+   catch(...)
+   {
+      _exception = std::current_exception();
+   }
+
+   result.trapped = !!_exception;
+   return result;
+}
+
+FizzyExecutionResult fizzy_runner::_invoke_system_call( const FizzyValue* args, FizzyExecutionContext* fizzy_context ) noexcept
+{
+   FizzyExecutionResult result;
+   result.has_value = false;
+   result.value.i64 = 0;
+
+   _exception = std::exception_ptr();
+
+   try
+   {
+      uint32_t xid = args[0].i32;
+      uint32_t ret_len = args[2].i32;
+      char* ret_ptr = resolve_ptr(_instance, args[1].i32, ret_len);
+      uint32_t arg_len = args[4].i32;
+      const char* arg_ptr = resolve_ptr(_instance, args[3].i32, arg_len);
+
+      KOINOS_ASSERT( ret_ptr != nullptr, wasm_memory_exception, "Invalid ret_ptr in invoke_thunk()" );
```

Comment:
In error message: `invoke_thunk()` should be `invoke_system_call()`.

Reply:
Good catch, it was a copy-paste error.